pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -326,14 +326,27 @@ func (p *staticPolicy) allocateCPUs(s state.State, numCPUs int, numaAffinity bit
 
 	allocatableCPUs := p.GetAllocatableCPUs(s).Union(reusableCPUs)
 
-	// If there are aligned CPUs in numaAffinity, attempt to take those first.
+	// re-use CPUs first: init-containers CPUs are reserved for the current pod
 	result := cpuset.NewCPUSet()
+	if reusableCPUs.Size() > 0 {
+		neededReusableCPUs := reusableCPUs.Size()
+		if numCPUs < neededReusableCPUs {
+			neededReusableCPUs = numCPUs
+		}
+		var err error
+		result, err = p.takeByTopology(reusableCPUs, neededReusableCPUs)
+		if err != nil {
+			return cpuset.NewCPUSet(), err
+		}
+	}
+
+	// If there are aligned CPUs in numaAffinity, attempt to take those next.
```

Comment:
What if the reusable CPUs are not aligned with the `TopologyHint`? Even though the init CPUs are reusable, we only want to actually reuse them if they satisfy the same NUMA constraints as the app container.

The current implementation will consider the reusable CPUs as allocatable in this line:
```
allocatableCPUs := p.GetAllocatableCPUs(s).Union(reusableCPUs)
```
And then grab the "alignable" ones from that set before making an allocation decision.

What seems to be missing, is to _prefer_ alignable CPUs from the reusable set over other CPUs.

So while I don't think the change you propose will work as is, we should definitely put some logic in there to make the flow as follows:
1. Pull as many aligned CPUs form the reusable set as possible
2. Pull as many CPUs from the remaining aligned CPU set as possible
3. Pull as many non-aligned CPUs from the reusable set as required
4. Pull as many non-aligned CPUs from the remaining CPU set as required

While this doesn't guarantee that all `reusable` CPUs will get reused, it does ensure that the alignment guarantees of the `ToplogyManager` are properly enforced, while reusing as many reusable CPUs as possible.

In an ideal world we would then allow the non-reused CPU to be freed after the init container that used it completed. This is a bigger change though, and one that already has an outstanding issue created for it: https://github.com/kubernetes/kubernetes/issues/94220

Reply:
Noted, thanks. I assumed that alignement was always best-effort rather than guaranteed but now I understand it's not ; so we don't have options to guarantee not leaking CPUs until we support freeing init-containers resources?

Reply:
The only way to guarantee you do not leak them without this freeing mechanism, is to set up your resource request in the init container to match the resource request in your main container (even if your init container doesn't actually need all of the resources that the main container does). It's a bit hacky since it relies on underlying implementation details, but it will guarantee that the same set of CPUs and devices are considered in the same order so that the exact same allocation decision is made.

Reply:
We try this workaround, but it failed, even make more leaking init CPUs. It happened in kubernetes 1.19.7.
check pod d18dd176-f8be-48e3-8c63-c6c0f7ede780 below:
```
{
   "policyName":"static",
   "defaultCpuSet":"0,2,84-87",
   "entries":{
      "0062ee26-d0c1-4437-900e-439ae2db073d":{
         "app":"22-25,96-107",
         "init-container":"1"
      },
      "4d88429b-8a78-47f8-8a8a-550af683d9b3":{
         "app":"14-21,88-95",
         "init-container":"88"
      },
      "6252c71a-8b3d-4567-8869-73932f4e4583":{
         "app":"4-7,38-49",
         "init-container":"3"
      },
      "beb70340-0bbb-4fa8-a51f-af303d4044b8":{
         "app":"8-13,26-27,32-37,50-51",
         "init-container":"8"
      },
      "d18dd176-f8be-48e3-8c63-c6c0f7ede780":{
         "app":"108-123",
         "init-container":"64-65,70-83"
      },
      "eb6773bb-0c84-4517-86fc-73a5dc818a02":{
         "app":"28-31,66-69",
         "init-container":"28"
      },
      "eeeeddd9-eecd-47f4-8a0d-83e19063380c":{
         "app":"52-63,124-127",
         "init-container":"52-63,124-127"
      }
   },
   "checksum":2581470015
}
```

Reply:
From what I can tell in your dump of the checkpoint file, most of your pods don't have an equal number of CPUs requested by the init container and the app container (most of them only seem to have 1 CPU requested for the init container). That is what is required for this workaround.

The only one that does have an equal number of CPUs between the init container and the app container is the last one, and in that one we see all of them being reused with no leakage.

Reply:
please check pod `d18dd176-f8be-48e3-8c63-c6c0f7ede780`, that init-container and app-container got the equal number CPUs 16. However, they used completely different CPUset, '108-123' vs '64-65,70-83'. It means there are 16 leaking CPUs.

By the way, we are on the way of trying the workaround, only new pods(d18dd176-f8be-48e3-8c63-c6c0f7ede780 & eeeeddd9-eecd-47f4-8a0d-83e19063380c ) are set init-container euql to app-contaienr, that's why most of they have only 1 CPU requested for init container.